File: internal/namespace-semantics.ts
import { isNode as isAstNode, type AstNode, type TsModuleDeclaration } from "./ast.ts";

interface NamespaceSemanticNode extends AstNode {
	readonly body?: NamespaceSemanticNode | readonly NamespaceSemanticNode[] | null;
	readonly declaration?: NamespaceSemanticNode | null;
	readonly declare?: boolean;
	readonly exportKind?: "type" | "value";
	readonly importKind?: "type" | "value";
	readonly isExport?: boolean;
	readonly isTypeOnly?: boolean;
}

const moduleRuntimeState = new WeakMap<AstNode, boolean>();

export function isTypeOnlyModule(node: AstNode): boolean {
	if ((node as NamespaceSemanticNode).declare === true) {
		return true;
	}
	const cached = moduleRuntimeState.get(node);
	if (cached !== undefined) {
		return !cached;
	}
	const hasRuntimeState = moduleHasRuntimeState(node as NamespaceSemanticNode);
	moduleRuntimeState.set(node, hasRuntimeState);
	return !hasRuntimeState;
}

export function nearestRuntimeNamespace(ancestors: readonly AstNode[]): TsModuleDeclaration | null {
	for (let index = ancestors.length - 1; index >= 0; index -= 1) {
		const ancestor = ancestors[index]!;
		if (ancestor.type === "TSModuleDeclaration" && !isTypeOnlyModule(ancestor)) {
			return ancestor as TsModuleDeclaration;
		}
	}
	return null;
}

export function isTypeOnlyNamespaceExportDeclaration(node: AstNode): boolean {
	return (
		node.type === "TSInterfaceDeclaration" ||
		node.type === "TSTypeAliasDeclaration" ||
		node.type === "TSDeclareFunction" ||
		(node.type === "TSModuleDeclaration" && isTypeOnlyModule(node)) ||
		(node as NamespaceSemanticNode).declare === true
	);
}

export function isSupportedRuntimeNamespaceExportDeclaration(node: AstNode): boolean {
	if (node.type === "FunctionDeclaration" || node.type === "ClassDeclaration" || node.type === "TSEnumDeclaration") {
		return (node as NamespaceSemanticNode).declare !== true;
	}
	return node.type === "TSModuleDeclaration" && !isTypeOnlyModule(node);
}

function moduleHasRuntimeState(node: NamespaceSemanticNode): boolean {
	const body = node.body;
	if (!isAstNode(body)) {
		return true;
	}
	if (body.type === "TSModuleDeclaration") {
		return !isTypeOnlyModule(body);
	}
	if (body.type !== "TSModuleBlock" || !Array.isArray(body.body)) {
		return true;
	}
	return body.body.some((statement) => isAstNode(statement) && namespaceStatementHasRuntimeState(statement));
}

function namespaceStatementHasRuntimeState(node: NamespaceSemanticNode): boolean {
	if (
		node.type === "TSInterfaceDeclaration" ||
		node.type === "TSTypeAliasDeclaration" ||
		node.type === "TSNamespaceExportDeclaration"
	) {
		return false;
	}
	if (node.type === "TSImportEqualsDeclaration") {
		return !isNonInstantiatingNamespaceImport(node);
	}
	if (node.type === "TSModuleDeclaration") {
		return !isTypeOnlyModule(node);
	}
	if (node.type !== "ExportNamedDeclaration") {
		return true;
	}
	const declaration = node.declaration;
	if (isAstNode(declaration)) {
		return namespaceStatementHasRuntimeState(declaration);
	}
	return node.exportKind !== "type";
}

function isNonInstantiatingNamespaceImport(node: AstNode): boolean {
	if (node.type !== "TSImportEqualsDeclaration") {
		return false;
	}
	if ((node as NamespaceSemanticNode).importKind === "type" || (node as NamespaceSemanticNode).isTypeOnly === true) {
		return true;
	}

	// A namespace-local import alias does not create an exported value on its
	// own. If every other element in the namespace is type-only, the namespace
	// has no runtime instantiation and the alias disappears with that scope.
	// `export import` is different: it contributes a runtime property and must
	// keep the namespace instantiated.
	return (node as NamespaceSemanticNode).isExport !== true;
}

File: src/transform/namespace_semantics.zig
const parser = @import("parser");

const NodeIndex = parser.ast.NodeIndex;

/// Runtime-state classification shared by fixed erasure and namespace
/// lowering. It operates directly on Yuku nodes and intentionally does not add
/// an independent semantic validation layer.
pub fn is_type_only_module(
    tree: *const parser.ast.Tree,
    declaration: parser.ast.TSModuleDeclaration,
) bool {
    if (declaration.declare) return true;
    if (declaration.body == .null) return false;

    const block = switch (tree.data(declaration.body)) {
        .ts_module_block => |value| value,
        else => return false,
    };
    for (tree.extra(block.body)) |statement| {
        if (statement_has_runtime_state(tree, statement)) return false;
    }
    return true;
}

pub fn statement_has_runtime_state(tree: *const parser.ast.Tree, index: NodeIndex) bool {
    return switch (tree.data(index)) {
        .ts_interface_declaration,
        .ts_type_alias_declaration,
        .ts_namespace_export_declaration,
        .ts_global_declaration,
        => false,
        // A namespace-local alias does not instantiate the namespace by itself.
        .ts_import_equals_declaration => false,
        .ts_module_declaration => |node| !is_type_only_module(tree, node),
        .export_named_declaration => |node| if (node.declaration != .null)
            switch (tree.data(node.declaration)) {
                .ts_import_equals_declaration => |declaration| declaration.import_kind != .type,
                else => statement_has_runtime_state(tree, node.declaration),
            }
        else
            node.export_kind != .type,
        else => true,
    };
}

pub fn is_supported_runtime_export_declaration(
    tree: *const parser.ast.Tree,
    index: NodeIndex,
) bool {
    return switch (tree.data(index)) {
        .function => |node| node.type == .function_declaration and !node.declare,
        .class => |node| node.type == .class_declaration and !node.declare,
        .ts_enum_declaration => |node| !node.declare,
        .ts_module_declaration => |node| !is_type_only_module(tree, node),
        else => false,
    };
}
